Add then(callback)->Future API to ivalue::Future

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37311 Add then(callback)->Future API to utils/Future.h**



Differential Revision: [D21247827](https://our.internmc.facebook.com/intern/diff/D21247827)

Commit:
Let rpc.Future expose an add_done_callback API

[ghstack-poisoned]

Commit:
Update on "Let rpc.Future expose an add_done_callback API"


Differential Revision: [D21247827](https://our.internmc.facebook.com/intern/diff/D21247827)

[ghstack-poisoned]

Commit:
Update on "Let rpc.Future expose an add_done_callback API"


Differential Revision: [D21247827](https://our.internmc.facebook.com/intern/diff/D21247827)

[ghstack-poisoned]

Commit:
Update on "Let rpc.Future expose an add_done_callback API"


Differential Revision: [D21247827](https://our.internmc.facebook.com/intern/diff/D21247827)

[ghstack-poisoned]

Commit:
Update on "Let rpc.Future expose an add_done_callback API"


Differential Revision: [D21247827](https://our.internmc.facebook.com/intern/diff/D21247827)

[ghstack-poisoned]

Commit:
Update on "Let rpc.Future expose an add_done_callback API"


Differential Revision: [D21247827](https://our.internmc.facebook.com/intern/diff/D21247827)

[ghstack-poisoned]

pritamdamania87 commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -2355,6 +2367,71 @@ def test_non_cont_tensors(self):
             self.assertEqual(t_view, t_ret)
             self.assertFalse(t_ret.is_contiguous())
 
+    @dist_init
+    def test_callback(self):
```
Can we add a few more test cases:

1) Test that the callback can be a torchscript function.
2) Attach python callback to future returned by:
     i) rpc_async within torchscript.
     ii) rpc_async calling torchscript function.
3) Add tests to validate that we receive appropriate exceptions if callback doesn't have appropriate signature.

mrshenli replied:
Cannot test calling `_then` in a TorchScript function yet, as that would require making `_then` torchscrptable.

pritamdamania87 commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -331,23 +344,68 @@ PyObject* rpc_init(PyObject* /* unused */) {
+          )")
+      .def(
+          "add_done_callback",
+          [&](FutureIValue& fut, py::function cb) {
```
Should we have some validation here to ensure that cb is valid? For example its not None and it has the correct signature?

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -23,6 +24,18 @@ namespace rpc {
 namespace {
 
+// Wrap Python function to guard deref
+struct PythonFunction {
```
Will rebase when #35154 is landed to see if we still need this.

Commit:
Update on "Let rpc.Future expose an add_done_callback API"


Differential Revision: [D21247827](https://our.internmc.facebook.com/intern/diff/D21247827)

[ghstack-poisoned]

Commit:
Update on "Add then(callback)->Future API to utils/Future.h"




Differential Revision: [D21247827](https://our.internmc.facebook.com/intern/diff/D21247827)

[ghstack-poisoned]

mrshenli commented on torch/csrc/utils/future.h:
```diff
@@ -133,6 +133,19 @@ class TORCH_API Future final {
   }
 
+  template<typename R>
+  std::shared_ptr<Future<R>> then(std::function<R(const Future<T>& future)> cb) {
```
I am not sure if this is the appropriate API. Since the returned Future represents the completion of the callback, it should also hold the returned value of the callback?

rohan-varma replied:
What's the reason for the CB taking in a future instead of the completed value of the future? Is it so that we can access errors on the future when running the callback?

mrshenli replied:
> what's the reason for the CB taking in a future instead of the completed value of the future? Is it so that we can access errors on the future when running the callback (if so see above comment)?

Yep, the return value `T` might not be available if the callback crashed.

Commit:
Update on "Add then(callback)->Future API to utils/Future.h"




Differential Revision: [D21247827](https://our.internmc.facebook.com/intern/diff/D21247827)

[ghstack-poisoned]

Commit:
Update on "Add then(callback)->Future API to utils/Future.h"




Differential Revision: [D21247827](https://our.internmc.facebook.com/intern/diff/D21247827)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -333,23 +346,71 @@ PyObject* rpc_init(PyObject* /* unused */) {
+            // here to explicitly acquire GIL and decouple value destruction and
+            // callback destruction.
+            PythonFunction pf(std::move(cb));
+            return fut.then<IValue>([pf](const FutureIValue& fut) -> IValue {
```
nit: I think the `const FutureIValue& fut` shadows the self argument `FutureIValue& fut` on 367?

mrshenli replied:
ahh, right, let me change it to a different name.

rohan-varma commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -333,23 +346,71 @@ PyObject* rpc_init(PyObject* /* unused */) {
+                  py::object ret =
+                      pf.func_(torch::jit::toPyObject(fut.constValue()));
+                  return jit::toIValue(std::move(ret), PyObjectType::get());
+                } catch (py::error_already_set& e) {
```
Thanks for taking care of the error handling! This is great, though I was wondering what you think about making the python user's cb signature something like:
```
def callback(future_result, future_error):
    # User's responsibility to check and handle error
```

with this, we could avoid passing in the future here and instead pass in the result and optional error.

Of course, this is also fine, and we can revisit if user code needs custom error handling.

mrshenli replied:
> For example, there may be cases where the user does not want to throw if the parent future failed.

I see, this makes sense. In this case, should we let the Python callback API also takes the original future instead of the return value? Otherwise, the two args will be mutual exclusive optional return value and optional error?

rohan-varma replied:
This would be good too, although it can be done as a follow up too since I don't feel there's a need to block this diff on that. The API is private for now as well.

Commit:
Update on "Add then(callback)->Future API to utils/Future.h"




Differential Revision: [D21247827](https://our.internmc.facebook.com/intern/diff/D21247827)

[ghstack-poisoned]

Commit:
Update on "Add then(callback)->Future API to utils/Future.h"




Differential Revision: [D21247827](https://our.internmc.facebook.com/intern/diff/D21247827)

[ghstack-poisoned]

rohan-varma commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -2355,6 +2367,122 @@ def test_non_cont_tensors(self):
+
+        self.assertEqual(fut.wait(), torch.ones(n, n) * 2)
+        self.assertEqual(set_by_cb.result(), torch.ones(n, n) * 2 + 1)
+        self.assertEqual(fut.wait(), torch.ones(n, n) * 2)
```
Is this being here twice intentional? If so, what's it testing?

mrshenli replied:
This is testing the cb didn't mess up with the value in the Future. Let me add a comment

rohan-varma commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -2355,6 +2367,122 @@ def test_non_cont_tensors(self):
+
+        ret = rpc.rpc_sync(
+            dst1,
+            add_use_future_cb,
```
Nice! Is there a test where a user-defined cb runs on the remote side, throws, and we ensure the right error is propagated back to the client? (I saw there is an error check test above where the cb is local)

mrshenli replied:
Let me add this test as well.

rohan-varma reviewed (approved):
Thanks for getting this together so quickly! Looks good to me % some comments. Also it looks like @pritamdamania87 has some comments as well

Commit:
Update on "Add then(callback)->Future API to utils/Future.h"




Differential Revision: [D21247827](https://our.internmc.facebook.com/intern/diff/D21247827)

[ghstack-poisoned]

Commit:
Update on "Add then(callback)->Future API to utils/Future.h"




Differential Revision: [D21247827](https://our.internmc.facebook.com/intern/diff/D21247827)

[ghstack-poisoned]

Commit:
Update on "Add then(callback)->Future API to utils/Future.h"




Differential Revision: [D21247827](https://our.internmc.facebook.com/intern/diff/D21247827)

[ghstack-poisoned]

mrshenli committed:
Update on "Add then(callback)->Future API to utils/Future.h"




Differential Revision: [D21247827](https://our.internmc.facebook.com/intern/diff/D21247827)

[ghstack-poisoned]

Commit:
Update on "Add then(callback)->Future API to utils/Future.h"




Differential Revision: [D21247827](https://our.internmc.facebook.com/intern/diff/D21247827)

[ghstack-poisoned]

Commit:
Update on "Add then(callback)->Future API to utils/Future.h"




Differential Revision: [D21247827](https://our.internmc.facebook.com/intern/diff/D21247827)

[ghstack-poisoned]

Commit:
Update on "Add then(callback)->Future API to utils/Future.h"




Differential Revision: [D21247827](https://our.internmc.facebook.com/intern/diff/D21247827)

[ghstack-poisoned]

Commit:
Update on "Add then(callback)->Future API to utils/Future.h"




Differential Revision: [D21247827](https://our.internmc.facebook.com/intern/diff/D21247827)

[ghstack-poisoned]

Commit:
Update on "Add then(callback)->Future API to utils/Future.h"




Differential Revision: [D21247827](https://our.internmc.facebook.com/intern/diff/D21247827)

[ghstack-poisoned]

Commit:
Update on "Add then(callback)->Future API to utils/Future.h"




Differential Revision: [D21247827](https://our.internmc.facebook.com/intern/diff/D21247827)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -86,10 +95,67 @@ struct VISIBILITY_HIDDEN PythonFutureWrapper {
+        [pyFut(this->getPtr()), pf(std::move(pf))]() -> IValue {
+          try {
+            pybind11::gil_scoped_acquire ag;
+            return toIValue(pf->func_(pyFut), PyObjectType::get());
```
Do we need to do any validation on `func_` having the right signature or is the error message when the args are incorrect okay?

mrshenli replied:
pybind covers this, there is a `test_callback_wrong_args` test to make sure the error message is clear.

rohan-varma commented:
LGTM, did some testing and verified that I can use this to fix the profiler test issue.

pritamdamania87 commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -828,6 +828,10 @@ void initJITBindings(PyObject* module) {
       .def(
           "wait",
           &PythonFutureWrapper::wait,
+          py::call_guard<py::gil_scoped_release>())
+      .def(
+          "_then",
+          &PythonFutureWrapper::then,
```
Is RPC and IValue future already merged? :)

mrshenli replied:
Yes, they are. See #35154

pritamdamania87 commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -86,10 +95,67 @@ struct VISIBILITY_HIDDEN PythonFutureWrapper {
+            }
+
+            throw err;
+          } catch (...) {
+            throw std::runtime_error("Unknown error when running callback");
+          }
```
Do we need to catch and rethrow this? Can't we just avoid catching in the first place? That would also preserve the original exception.

pritamdamania87 commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -86,10 +95,67 @@ struct VISIBILITY_HIDDEN PythonFutureWrapper {
+            pybind11::gil_scoped_acquire ag;
+            return toIValue(pf->func_(pyFut), PyObjectType::get());
+          } catch (py::error_already_set& e) {
+            auto err = std::runtime_error(c10::str(
+                "Got the following error when running the callback: ",
+                e.what()));
+            {
+              pybind11::gil_scoped_acquire ag;
+              // Release ownership on py::objects and also restore Python
+              // Error Indicator.
+              e.restore();
+              // Clear the Python Error Indicator as we has recorded the
+              // exception in the response message.
+              PyErr_Clear();
+            }
+
+            throw err;
+          } catch (...) {
+            throw std::runtime_error("Unknown error when running callback");
```
Shouldn't we mark the future that we're returning as part of `then` with these errors instead of just rethrowing them?

mrshenli replied:
`ivalue::Future::then(cb)` did that.

pritamdamania87 commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -828,6 +828,10 @@ void initJITBindings(PyObject* module) {
           &PythonFutureWrapper::wait,
+          py::call_guard<py::gil_scoped_release>())
+      .def(
+          "_then",
```
Why `_then` instead of just `then`? Is this some convention in JIT?

mrshenli replied:
No, not sure if the current signature is appropriate, but need to get this in earlier to unblock fixing profiler tests. So mark it `_then` with no API doc for now, and will create a PR on top to expose this API publicly.

pritamdamania87 commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -2366,6 +2381,154 @@ def test_non_cont_tensors(self):
+        self.assertEqual(fut.wait(), torch.ones(n, n) * 2)
+
+    @dist_init
+    def test_callback_chain(self):
```
Can we have a test similar to this where the remote function is a torchscript function?

mrshenli replied:
sure adding.

pritamdamania87 reviewed (commented):
Have a few comments inline. Also, was wondering if we should get a review from @wanchaol for the JIT changes?

Commit:
Update on "Add then(callback)->Future API to utils/Future.h"




Differential Revision: [D21247827](https://our.internmc.facebook.com/intern/diff/D21247827)

[ghstack-poisoned]

wanchaol commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -86,10 +95,65 @@ struct VISIBILITY_HIDDEN PythonFutureWrapper {
+        [pyFut(this->getPtr()), pf(std::move(pf))]() -> IValue {
+          try {
+            pybind11::gil_scoped_acquire ag;
+            return toIValue(pf->func_(pyFut), PyObjectType::get());
```
does this mean that the py::function passed in should always require the user to pass a  future? shall we document this?

mrshenli replied:
Yes, let me add a comment here

wanchaol reviewed (approved):
looks good to me.

Commit:
Update on "Add then(callback)->Future API to utils/Future.h"




Differential Revision: [D21247827](https://our.internmc.facebook.com/intern/diff/D21247827)

[ghstack-poisoned]
